fix(semantic): handle edge cases checking `super`

overlookmotel opened the pull request:
Fixes #13284.

Fix semantic's checking of `super()` and `super.prop`.

This fixes various TS test cases, notably correctly identifying that these are all illegal:

```js
class C {
  prop = function() { super.foo };
  [super.foo]() {}
  @super.foo method() {}
};
```

It also fixes various edge cases related to classes nested within other classes that aren't covered by existing conformance tests.

It's pretty much a complete re-write. I've combined the 2 paths for in a class and not in a class into one, to avoid repeated code.

This PR includes pass + fail test fixtures which include `super` in every conceivable position I could think of. There are so many errors in the fail test fixtures, that I had to write a script to check the snapshot includes them all!

Hopefully this solves this issue for once and for all!

Note: The large pass fixture seems to be revealing bugs somewhere in transformer - there's a mismatch, as well as lots of wrong scopes. But that's incidental to this PR, and can be addressed separately.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13499?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#13499** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13499?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13499?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#13463** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13463?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13499).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_semantic/src/checker/javascript.rs:
```diff
@@ -891,110 +892,239 @@ fn unexpected_super_reference(span: Span) -> OxcDiagnostic {
-            ctx.error(unexpected_super_call(super_call_span));
-        } else {
-            ctx.error(unexpected_super_reference(sup.span));
+            unreachable!();
```
There is a lengthy comment explaining exactly why it's not possible for the loop to terminate without matching one of the expected ancestor nodes. If there *is* a mistake in the logic, we want a panic, so we find out about it.

overlookmotel commented on crates/oxc_semantic/src/checker/javascript.rs:
```diff
@@ -891,110 +892,239 @@ fn unexpected_super_reference(span: Span) -> OxcDiagnostic {
-            ctx.error(unexpected_super_call(super_call_span));
-        } else {
-            ctx.error(unexpected_super_reference(sup.span));
+            unreachable!();
```
See my reply to Copilot above who said the same thing.

overlookmotel commented:
No new errors in `monitor-oxc`, so `Super` in `TSIndexSignature` isn't in any of those examples. But I've added a commit which handles it, to be on safe side.

Dunqing commented:
I've looked into the mismatch case, which is related to the legacy decorator, and it probably doesn't matter because it only affects class expressions, but the legacy decorator doesn't support transforming class expressions with decorators.

Dunqing commented:
### Merge activity

* **Sep 10, 3:23 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(semantic): handle edge cases checking `super` (#13499)

Fixes #13284.

Fix semantic's checking of `super()` and `super.prop`.

This fixes various TS test cases, notably correctly identifying that these are all illegal:

```js
class C {
  prop = function() { super.foo };
  [super.foo]() {}
  @super.foo method() {}
};
```

It also fixes various edge cases related to classes nested within other classes that aren't covered by existing conformance tests.

It's pretty much a complete re-write. I've combined the 2 paths for in a class and not in a class into one, to avoid repeated code.

This PR includes pass + fail test fixtures which include `super` in every conceivable position I could think of. There are so many errors in the fail test fixtures, that I had to write a script to check the snapshot includes them all!

Hopefully this solves this issue for once and for all!

Note: The large pass fixture seems to be revealing bugs somewhere in transformer - there's a mismatch, as well as lots of wrong scopes. But that's incidental to this PR, and can be addressed separately.
